### DISCUSSION TITLE
At the same time meet the dimension and a priori

### DISCUSSION THREAD

**Discussion by jackeuylov** [MilesCranmer/PySR#664] [Q&A]
Hi Miles,
In my research, I need to make some kind of a priori of the equation form, and at the same time, I need to satisfy the balance of dimensions. However, when I custom-designed the loss, I found that the penalty of the dimensional part was offset. Here is my code example :
```python
x = np.random.rand(1000,3)
ones_array = np.ones((1000, 1))
X = np.hstack((x, ones_array))
y = X[:, 0]/(X[:, 1]*X[:, 2])+X[:, 1]/(X[:, 0]*X[:, 0]*X[:, 3])

objective = """
function my_custom_objective(tree, dataset::Dataset{T}, options) where {T<:Real}
.....my loss.....(not contain dimension)
"""
model = PySRRegressor(
    population_size=50,
    ncyclesperiteration=500,
    niterations=800,
    early_stop_condition=("stop_if(loss, complexity) = loss < 1e-5"),
    binary_operators=["+","-","*","/"],
    unary_operators=["square"],
    complexity_of_constants=100,
    maxsize=40,
    maxdepth=40,    
    loss_function=objective,
    dimensional_constraint_penalty=10**4,
)
model.fit(X, y, X_units=["Pa", "Pa", "1", "1/Pa"],y_units=["1"])
```
I found that the above loss does not deal with the problem of dimension.
Could you help me see how to add dimensional constraints to custom loss ?
Looking forward to your reply! Thanks.

**Comment by MilesCranmer** [comment]
The default loss is here: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/cd23a6e25c64d00565c3ae3905d06dc3c63033ed/src/LossFunctions.jl#L45-L75

In particular the dimensional regularization is calculated on this line: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/cd23a6e25c64d00565c3ae3905d06dc3c63033ed/src/LossFunctions.jl#L71

So within a custom loss you could add a manual call, like

```julia
loss_val += SymbolicRegression.LossFunctionsModule.dimensional_regularization(tree, dataset, options)
```
